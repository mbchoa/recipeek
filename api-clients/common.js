import t from 'tcomb';

const Method = t.enums.of(['GET', 'POST', 'HEAD', 'PUT', 'DELETE'], 'Method');
const Headers = t.dict(t.String, t.String);

export const RequestOptions = t.struct({
  method: Method,
  headers: Headers,
  url: t.String,
}, 'RequestOptions');
