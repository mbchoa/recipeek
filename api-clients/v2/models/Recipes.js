import t from 'tcomb';
import { map } from 'lodash';

import Recipe from './Recipe';

const Recipes = t.struct({
  endIndex: t.Number,
  recipeList: t.list(Recipe),
  startIndex: t.Number,
}, 'Recipes');

Recipes.of = t.func(
  [t.Object],
  Recipes,
  'Recipes'
).of(payload => {
  const {
    from: startIndex,
    hits,
    to: endIndex,
  } = payload;

  return {
    endIndex,
    recipeList: map(hits, recipe => Recipe.of(recipe)),
    startIndex,
  };
});

export default Recipes;
