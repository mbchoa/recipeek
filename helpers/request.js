import { identity, isError, omit } from 'lodash';
import fetch from 'whatwg-fetch';

function checkStatus(response) {
  const { status } = response;
  return status >= 200 && status < 300
    ? response
    : new Error(response);
}

export function makeRequest(
  request = fetch,
  options = {},
  createResult = identity
) {
  return new Promise((resolve, reject) => {
    request(options.url, omit(options, 'url'))
      .then(response => {
        let result;
        try {
          result = createResult(response.text());
        } catch (err) {
          return reject(err);
        }

        if (!isError(checkStatus(response))) {
          return resolve(result);
        }

        reject(new Error(`Unhandled status code: ${response.status} received.`));
      });
  });
}

export function createMockRequest() {
  const requests = [];

  function request(url, options) {
    let resolver;
    let rejector;
    const promise = new Promise((resolve, reject) => {
      resolver = resolve;
      rejector = reject;
    });

    requests.push({
      rejector,
      resolver,
    });

    return promise;
  }

  function validateResponse() {
    if (!requests.length) {
      throw new Error('A response was provided, but no request was expecting it');
    }
  }

  function decorateResponse(response) {
    return {
      text() {
        return response;
      },
      status: response.status,
    };
  }

  function resolveResponse(response) {
    validateResponse();
    requests
      .shift()
      .resolver(decorateResponse(response));
  }

  function rejectResponse(response) {
    validateResponse();
    requests
      .shift()
      .rejector(decorateResponse(response));
  }

  return { rejectResponse, request, resolveResponse };
}
