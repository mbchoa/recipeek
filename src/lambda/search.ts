import axios, { AxiosInstance } from 'axios';
import qs from 'qs';

const client: AxiosInstance = axios.create({
  baseURL: 'https://api.edamam.com',
  timeout: 5000
});

export async function handler(event: any) {
  const { query } = event.queryStringParameters;
  try {
    const { data } = await client.get(
      `/search${qs.stringify(
        {
          app_id: process.env.REACT_APP_RECIPEEK_APP_ID,
          app_key: process.env.REACT_APP_RECIPEEK_APP_KEY,
          q: query
        },
        { addQueryPrefix: true }
      )}`,
      { headers: { Accept: 'application/json' } }
    );
    return {
      statusCode: 200,
      body: JSON.stringify(data.hits)
    };
  } catch (err) {
    console.log(err); // output to netlify function log
    return {
      statusCode: 500,
      body: JSON.stringify({ msg: err.message }) // Could be a custom message or object i.e. JSON.stringify(err)
    };
  }
}
