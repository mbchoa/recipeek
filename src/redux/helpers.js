export const fsa = type => () => ({ type });
export const fsaPayload = type => payload => ({
  type,
  payload
});
