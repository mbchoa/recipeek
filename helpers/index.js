import request from './request';
import string from './string';

export default {
  request,
  string,
};
