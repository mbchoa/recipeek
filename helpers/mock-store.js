import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { createMockRequest } from './request';

const { request } = createMockRequest();

export default function mockStore(
  state = {},
  options = request
) {
  return configureStore([
    thunk.withExtraArgument(options),
  ])(state);
}
