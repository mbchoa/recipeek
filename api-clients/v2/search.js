import { attempt, isError } from 'lodash';
import t from 'tcomb';
import { NOT_FOUND, OK } from 'http-status-codes';

import { makeRequest } from '../../helpers/request';
import { RequestOptions } from '../common';
import Recipes from './models/Recipes';

export const createRequestOptions = t.func(
  [t.String],
  RequestOptions
).of(ingredient => ({
  method: 'GET',
  url: `/api/v2/search/${ingredient}`,
  headers: {
    'Content-Type': 'application/json; charset=utf-8',
    'Accept-Encoding': 'gzip'
  },
}));

export const createResult = ({ data, status }) => {
  switch (status) {
    case OK:
    case NOT_FOUND:
      return !isError(data)
        ? Recipes.of(data)
        : new Error('Error invalid data response received');

    default:
      return new Error(`Unhandled status code: ${status} received`);
  }
};

export default function search(request, ingredient) {
  return makeRequest(
    request,
    createRequestOptions(ingredient),
    createResult
  );
}
