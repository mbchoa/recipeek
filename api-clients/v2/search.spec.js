import { isError, noop, omit } from 'lodash';

import search, { createRequestOptions, createResult } from './search';
import { RequestOptions } from '../common';
import { createMockRequest } from '../../helpers/request';
import { recipesOk } from './fixtures';
import Recipes from './models/Recipes';

describe('v2 Search API Tests', () => {

  describe('#createRequestOptions', () => {

    it('creates request options with the correct defaults', () => {
      expect(RequestOptions.is(createRequestOptions('chicken')))
        .toBeTruthy();
    });

    it('should generate the correct URL request string', () => {
      expect(createRequestOptions('chicken').url)
        .toEqual('/api/v2/search/chicken');
    });

    it('should call request with requestOptions', () => {
      const request = jest.fn(() => new Promise(noop, noop));
      const requestOptions = createRequestOptions('chicken');

      search(request, 'chicken');

      expect(request.calls).toMatchSnapshot();
    });

  });

  describe('#createResult', () => {

    it('should return new Error object if data is an error object', () => {
      expect(isError(createResult({
        data: new Error(),
        status: 200,
      }))).toBeTruthy();
    });

    it('should return new Error object if status code received is unhandled', () => {
      expect(isError(createResult({
        data: { hello: 'world' },
        status: 500,
      }))).toBeTruthy();
    });

  });

  describe('#search', () => {
    const { request, resolveResponse } = createMockRequest();
    it('should resolve with payload with an OK response', done => {
      search(request, 'chicken')
        .then(result => {
          expect(Recipes.is(result))
            .toBeTruthy();
          done();
        }, done.fail);

      resolveResponse(recipesOk);
    });

  });
});
