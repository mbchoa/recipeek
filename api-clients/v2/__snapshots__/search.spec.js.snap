// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`v2 Search API Tests #createRequestOptions should call request with requestOptions 1`] = `undefined`;
